api/kv/value/value.go

```diff
@@ -130,6 +135,52 @@ func Uint(v uint) Value {
 	return Uint64(uint64(v))
 }
 
+func isAllowedType(object interface{}) bool {
+	switch reflect.TypeOf(object).Kind() {
+	case reflect.Array, reflect.Slice,
+		reflect.Float32, reflect.Float64,
+		reflect.Int, reflect.Int32, reflect.Int64,
+		reflect.Uint, reflect.Uint32, reflect.Uint64,
+		reflect.String:
+		return true
+	}
+	return false
+}
+
+// isArrayHomogenous checks if array elements are same type
+func isArrayHomogenous(collection interface{}) bool {
+	slice := reflect.ValueOf(collection)
+	sliceType := reflect.TypeOf(slice.Index(0).Interface())
+	kind := sliceType.Kind()
+	for i := 0; i < slice.Len(); i++ {
+		item := slice.Index(i).Interface()
+		if reflect.TypeOf(item).Kind() != kind ||
+			reflect.TypeOf(item) != sliceType ||
+			!isAllowedType(item) {
+			return false
+		}
+		if reflect.TypeOf(item).Kind() == reflect.Slice ||
+			reflect.TypeOf(item).Kind() == reflect.Array {
```

Comment:
Is this a recursive test? I do not believe the specification allows nested values in arrays at present. (I am advocating for any valid JSON value as an allowed value, which would let us throw away this homogeneity check.)